design-documents/jwt-support.md

```diff
@@ -0,0 +1,246 @@
+ - supports detached payload, multiple signatures, nested tokens
+
+## Implementation
+
+The following diagram represents needed interfaces to unify the workflow for JWT usage.
+
+![Class diagram](img/jwt-class-diagram.png)
+
+The `\Magento\Framework\Jwt\KeyGeneratorInterface` provides a possibility to use different types of key generators like: Magento deployment secret key, X.509 certificates or RSA keys.
+```php
+interface KeyGeneratorInterface
+{
+    public function create(): Jwk;
+}
+
+class CryptKey implements KeyGeneratorInterface
+{
+    public function __construct(SecretKeyFactory $keyFactory, DeploymentConfig $deploymentConfig)
+    {
+        $this->deploymentConfig = $deploymentConfig;
+        $this->keyFactory = $keyFactory;
+    }
+
+    public function create(): Jwk
+    {
+        $secret = (string) $this->deploymentConfig->get('crypt/key');
+        return $this->keyFactory->create($secret);
+    }
+}
+```
+
+The `\Magento\Framework\Jwt\ManagementInterface` is a base abstraction for JWT encoding/decoding/verification:
+```php
+interface ManagementInterface
+{
+    /**
+     * Generates JWT in header.payload.signature format.
+     *
+     * @param array $claims
+     * @return string
+     * @throws \Exception
+     */
+    public function encode(array $claims): string;
+
+    /**
+     * Parses JWT and returns payload.
+     *
+     * @param string $token
+     * @return array
+     */
+    public function decode(string $token): array;
+
+    /**
+     * Verifies JWT signature and claims.
+     * Throws \InvalidArgumentException if claims verification fails.
+     *
+     * @param string $token
+     * @return bool
+     * @throws \InvalidArgumentException
+     */
+    public function verify(string $token): bool;
+}
+```
+
+It has implementations for JWS and JWE. The default preference is JWS implementation and looks like this:
+```php
+class Management implements ManagementInterface
+{
+    public function __construct(
+        KeyGeneratorInterface $keyGenerator,
+        SerializerInterface $serializer,
+        AlgorithmFactory $algorithmFactory,
+        Json $json,
+        ClaimCheckerManager $claimCheckerManager
+    ) {
+        $this->keyGenerator = $keyGenerator;
+        $this->serializer = $serializer;
+        $this->algorithmFactory = $algorithmFactory;
+        $this->json = $json;
+        $this->claimCheckerManager = $claimCheckerManager;
+    }
+
+    public function encode(array $claims): string
+    {
+        // as payload represented by url encode64 on json string,
+        // the same claims structure with different key's order will get different payload hash
+        ksort($claims);
+        $payload = $this->json->serialize($claims);
+
+        $jwsBuilder = new JWSBuilder($this->algorithmFactory->getAlgorithmManager());
+        $jws = $jwsBuilder->create()
+            ->withPayload($payload)
+            ->addSignature(
+                $this->keyGenerator->create()->getKey(),
```

Comment:
`$this->keyGenerator->create()->getKey()` looks strange. I thought keys should be relatively stable, why do we need factory-like class here?

Reply:
The `\Magento\Framework\Jwt\KeyGeneratorInterface::create` returns the `\Magento\Framework\Jwt\Data\Jwk` which is a wrapper for key management from the 3rd party library. This wrapper allows avoiding the direct dependency from modules which might have own key creation mechanism, like CardinalCommerce, to the 3rd party library.